crates/primitives/receipt/src/from_blockifier.rs

```diff
@@ -82,8 +79,13 @@ pub fn from_blockifier_execution_info(res: &TransactionExecutionInfo, tx: &Trans
         _ => None,
     };
 
-    let messages_sent = res
-        .non_optional_call_infos()
+    fn recursive_call_info_iter(res: &TransactionExecutionInfo) -> impl Iterator<Item = &CallInfo> {
+        res
+            .non_optional_call_infos() // all root callinfos
+            .flat_map(|call_info| call_info.iter()) // flatmap over the roots' recursive inner call infos
+    }
+
+    let messages_sent = recursive_call_info_iter(res)
```

Comment:
why not do these 2 operations at the same time?

Reply:
sure okay

Reply:
it would be a bit messy with flat_map to unzip the iterator into two collections like this, do you have a suggestion that would look good?
i can't come up with a cleaner way than two iterations here and i don't think the perf implications are very important